File: src/aside/java/org/taktik/icure/logic/impl/MedicalLocationLogicImpl.java
package org.taktik.icure.logic.impl;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.taktik.icure.dao.MedicalLocationDAO;
import org.taktik.icure.dao.PlaceDAO;
import org.taktik.icure.entities.MedicalLocation;
import org.taktik.icure.entities.Place;
import org.taktik.icure.exceptions.DeletionException;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.logic.MedicalLocationLogic;

import java.util.List;

@Service
public class MedicalLocationLogicImpl extends GenericLogicImpl<MedicalLocation, MedicalLocationDAO> implements MedicalLocationLogic {

    private MedicalLocationDAO medicalLocationDAO;
    private ICureSessionLogic sessionLogic;

    @Override
    public MedicalLocation createMedicalLocation(MedicalLocation medicalLocation) {
        return medicalLocationDAO.create(medicalLocation);
    }

    @Override
    public List<String> deleteMedicalLocation(List<String> ids) throws DeletionException {
        try {
            deleteEntities(ids);
            return ids;
        } catch (Exception e) {
            throw new DeletionException(e.getMessage(), e);
        }
    }

    @Override
    public MedicalLocation getMedicalLocation(String medicalLocation) {
        return this.medicalLocationDAO.get(medicalLocation);
    }

    @Override
    public MedicalLocation modifyMedicalLocation(MedicalLocation medicalLocation) {
        return this.medicalLocationDAO.save(medicalLocation);
    }

    @Override
    public List<MedicalLocation> findByPostCode(String postCode) {
        return this.medicalLocationDAO.byPostCode(postCode);
    }

    @Override
    protected MedicalLocationDAO getGenericDAO() {
        return this.medicalLocationDAO;
    }

    @Autowired
    public void setSessionLogic(ICureSessionLogic sessionLogic) {
        this.sessionLogic = sessionLogic;
    }

    @Autowired
    public void setMedicalLocationDAO(MedicalLocationDAO dao) { this.medicalLocationDAO = dao; }
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/MedicalLocationLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import org.springframework.stereotype.Service
import org.taktik.couchdb.DocIdentifier
import org.taktik.icure.asyncdao.MedicalLocationDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.MedicalLocationLogic
import org.taktik.icure.entities.MedicalLocation
import org.taktik.icure.exceptions.DeletionException

@ExperimentalCoroutinesApi
@Service
class MedicalLocationLogicImpl(
	private val medicalLocationDAO: MedicalLocationDAO,
	private val sessionLogic: AsyncSessionLogic
) : GenericLogicImpl<MedicalLocation, MedicalLocationDAO>(sessionLogic), MedicalLocationLogic {

	override suspend fun createMedicalLocation(medicalLocation: MedicalLocation) = fix(medicalLocation) { medicalLocation ->
		medicalLocationDAO.create(medicalLocation)
	}

	override fun deleteMedicalLocations(ids: List<String>): Flow<DocIdentifier> {
		return try {
			deleteEntities(ids)
		} catch (e: Exception) {
			throw DeletionException(e.message, e)
		}
	}

	override suspend fun getMedicalLocation(medicalLocation: String): MedicalLocation? {
		return medicalLocationDAO.get(medicalLocation)
	}

	override suspend fun modifyMedicalLocation(medicalLocation: MedicalLocation) = fix(medicalLocation) { medicalLocation ->
		medicalLocationDAO.save(medicalLocation)
	}

	override fun findMedicalLocationByPostCode(postCode: String): Flow<MedicalLocation> = flow {
		emitAll(medicalLocationDAO.byPostCode(postCode))
	}

	override fun getGenericDAO(): MedicalLocationDAO {
		return medicalLocationDAO
	}
}
